### PR TITLE
Enable type-on-hover functionality in Vim

### PR DESCRIPTION
**Pull request by nilsbecker**
This PR refactors some functions in merlin.py to allow passing the mouse cursor position as input. In this way vim's balloonexpr functionality can be used to show the type at the mouse cursor. Hopefully this will be handy for browsing a codebase, where one does not necessarily wish to navigate the cursor to various places by keyboard commands.

Beware: the actualy popup window displayed by balloon-eval seems to be a bit flaky. I tested on two macos machines with identical MacVim version and identical .vimrc. The popup shows up only on one of the two. That is not a problem of merlin, as it fails also for the ALE plugin. The version with type display in the status line (or actually, below the status line? what is the bottom line of the screen called?) works in any case, so that should be a workable solution for everyone.

I'm happy for corrections, in particular also for the doc text.

### CODE REVIEW

File: vim/merlin/autoload/merlin.vim
Comment (trefis) [inline vim/merlin/autoload/merlin.vim:275]: I think I'd rather see the following function in `merlin.py`:
```python
def vim_type_enclosing_at_mouse():
  bufnr = vim.vvars['beval_bufnr']
  if bufnr != vim.current.buffer.number:
        return '{}'
  line = vim.vvars['beval_lnum']
  col = vim.vvars['beval_col']
  return type_enclosing_at_pos(line, col)
```

And then this line can be changed to
```
MerlinPy vim.command("let l:type = " + merlin.vim_type_enclosing_at_mouse())
```

Would there be an issue with that?

File: vim/merlin/autoload/merlin.vim
Comment (trefis) [inline vim/merlin/autoload/merlin.vim:279]: Can't this function be merged with the previous one?

File: vim/merlin/autoload/merlin.py
Comment (trefis) [inline vim/merlin/autoload/merlin.py:651]: This can disappear (and the function can be renamed to lose the `vim_` prefix) if the other change I proposed is implemented.

**Review by trefis — CHANGES_REQUESTED** [review]
Thanks for the PR!

I have a couple of suggestions, let me know if they make sense or not!

File: vim/merlin/autoload/merlin.vim
Comment (nilsbecker) [reply vim/merlin/autoload/merlin.vim:275]: That seems better, i'll try it.

**Review by trefis — APPROVED** [review]
LGTM.

**Comment by trefis** [comment]
(The CI failure seems unrelated, but I'll wait for @voodoos' confirmation before merging).

**Comment by voodoos** [comment]
> (The CI failure seems unrelated, but I'll wait for @voodoos' confirmation before merging).

Yes it was unrelated. @nilsbecker can you rebase your PR and add a changelog entry ?

**Comment by nilsbecker** [comment]
@voodoos does it look ok now?

File: CHANGES.md
Comment (voodoos) [inline CHANGES.md:15]:
```suggestion
    - add module, module type, and class imenu items for emacs (#1244, @ivg)
    - add type-on-hover functionality for vim (#1439)
```

**Comment by voodoos** [comment]
Thanks for the changes. It would be great if you could rebase to clean the history a bit by squashing the fixes and simplifications.

**Comment by voodoos** [comment]
Thank you !

**Pull request merged**
